src/groupeddataframe/grouping.jl

```diff
@@ -366,9 +378,33 @@ _nrow(x::NamedTuple{<:Any, <:Tuple{Vararg{AbstractVector}}}) =
 _ncol(df::AbstractDataFrame) = ncol(df)
 _ncol(x::Union{NamedTuple, DataFrameRow}) = length(x)
 
-function gen_fun(::Type{NT}, f2) where NT
-    get_val(::Val{T}) where {T} = T
+struct Reduce{F, A}
+    f::F
+    adjust::A
+end
+Reduce(f) = Reduce(f, nothing)
+function (r::Reduce)(x)
+    v = reduce(r.f, x)
+    r.adjust === nothing ? v : r.adjust(v, length(x))
```

Comment:
I would rather define it `r.adjust(v, x)` to allow any adjustment function. It is a bit more flexible then. Next, if we wanted even more generality (given my earlier comment what functions could be supported) - an even broader `Reduce` definition could be considered.

I would document and export `Reduce` so that users can write their custom reducers (e.g. geometric or harmonic mean).

Reply:
The problem is, the whole point of these optimizations is to avoid constructing the `x` vector. And even if we constructed it (using a view, it wouldn't be too costly, though probably significant), its length cannot be computed efficiently for `skipmissing` cases, while it's almost free to compute it on the fly. So for now I'd rather keep this internal, and discuss with e.g. SplitApplyCombine developers to see how this could be generalized. Special-casing `length` doesn't sound too bad since that's an essential information and it's very cheap to compute anyway.

Reply:
OK